compiler-cli/src/fs.rs

```diff
@@ -200,6 +200,21 @@ impl CommandExecutor for ProjectIO {
             }),
         }
     }
+
+    fn exists(&self, program: &str) -> Result<bool, Error> {
+        tracing::trace!(program=program, "command_exists");
+
+        match which::which(program) {
+            Ok(_) => Ok(true),
+            Err(which::Error::CannotFindBinaryPath) => Ok(false),
+            Err(err) => Err(Error::FileIo {
+                kind: FileKind::File,
+                action: FileIoAction::ReadMetadata,
+                path: Utf8PathBuf::from(program),
+                err: Some(err.to_string()),
+            }),
+        }
+    }
```

Comment:
Let's use the existing Elixir install locating code rather than adding this. The `maybe_link_elixir_libs` function links the Elixir libraries into place and errors if it is not possible.

Instead let's make it return whether it succeeded or not, and then have a "Elixir not installed" error be raised when a package is found that needs Elixir.

Reply:
Sorry, but I don't quite understand how this would be different from the original implementation, where the compiler checks if the dependency can be compiled with `mix`, and returns the same error from the shell like "Elixir not installed" (in my case asdf returns "No elixir executable found for elixir system")?

I reverted the last commit and checked it myself. 

1. When I have the `edgedb_gleam.ex` file in the `src/` directory:
```bash
❯ ../../gleam/target/debug/gleam build
  Resolving versions
Downloading packages
 Downloaded 3 packages in 0.01s
  Compiling gleam_stdlib
  Compiling gleeunit
  Compiling gleam_erlang
  Compiling edgedb_gleam
No elixir executable found for elixir system
error: Shell command failure

There was a problem when running the shell command `elixir`.
```
2. When I have the `edgedb` dependency depending on `crc` and `jose` dependencies, which in the case of elixir should be compiled with `mix`:
```bash
❯ ../../gleam/target/debug/gleam build
  Resolving versions
Downloading packages
 Downloaded 26 packages in 0.15s
  Compiling jose
No elixir executable found for elixir system
error: Shell command failure

There was a problem when running the shell command `elixir`.
```
3. When there is a dependency only on `crc`, but no need to call `mix`:
```bash
❯ ../../gleam/target/debug/gleam build
  Resolving versions
Downloading packages
 Downloaded 6 packages in 0.02s
  Compiling castore
  Compiling elixir_make
  Compiling gleam_stdlib
  Compiling gleeunit
  Compiling crc
 C      checksum_xor.c
 C      crc_8.c
 C      crc_model.c
 C      crc_algorithm.c
 C      crc_resource.c
 C      crc_nif.c
 C      xnif_slice.c
 LD     crc_nif.so
  Compiling gleam_erlang
  Compiling edgedb_gleam
   Compiled in 3.31s
```

Reply:
We want to use mix to compile a dep that can be compiled with mix or rebar if mix is installed.

We don't want to add new code for determining whether mix is installed, we already have code for that, the function `maybe_link_elixir_libs`. That should be used instead.

Reply:
I couldn't figure out exactly how to use the `maybe_link_elixir_libs` function in this context, but it looks like the new code does a similar thing as described above. 

It looks to see if the dependency can be built with `mix` and `rebar3` at the same time. And in that case, it tries at first to build with `mix`, which uses `maybe_link_elixir_libs` under the hood. And if `mix` compilation fails, it goes next with compiling the package with `rebar3` instead of returning an error. 

I hope this behavior is the expected one

Reply:
The `maybe_link_elixir_libs` returns an error when it cannot link Elixir into place. Rather than raise the error it should be recorded and then we can use that to determine if Elixir can be used, and only raise the error if Elixir is required.

I'm afraid I will not merge the PR as it stands as it introduces a second redundant way that we check for Elixir, and it adds a new unneeded library.

Reply:
Sorry, I did not know you had changed the code. Thank you